[Weave] signals content-DOCS-1994 (#2367)
wandb/docs: weave/guides/evaluation/monitors.mdx

---
title: "Monitor using built-in signals"
description: "W&B provides built-in scoring solutions for agents in production"
---

> In modern agent development, standard system metrics like latency, token count, and cost are insufficient for understanding complex agent behavior. While inspecting individual traces provides deep insight, it is impossible to scale manual reviews across the millions of traces generated in a live environment.

Reviewer: I think we should plainly state what signals are at the start. Otherwise, this kinda reads like a company blog post. We could amend this quickly by moving the second paragraph to the top and fixing it up a little at the end. We should also make the connection that signals are a type of monitor.

Also, I _feel_ like this paragraph is very opinionated. I think we should consider nixing it entirely and in lieu of something that plainly states what Signals do and how they benefit the customer.

Signals provide a high-level monitoring solution designed to bridge this gap by offering automated, behavioral scoring for agents in production.

> Signals utilize a robust backend infrastructure to provide real-time performance insights:

Reviewer: I'm skeptical about mentioning a backend infrastructure if customers don't have to manage/configure a backend infrastructure. 

Also, I'm so sorry, but my old manager is yelling at me in my head about never using the word "utilize" when you could write "use" 😅

> - Automated scoring: Every incoming production trace is automatically processed and scored based on predefined metrics.

Reviewer: Should these opening terms be bolded?

- Infrastructure: Processing is powered by Coreweave compute and Coreweave GPUs to ensure scalability across millions of traces.

> - Custom metrics: Developers can create specific metrics, such as response length or faithfulness to source material, to help understand exactly how an agent is behaving.

Reviewer: I think we might need to rewrite the opening line a little bit, because it makes it sound like this is going to be a list of performance metrics that it can return, but these are really features of the feature.

By using signals within production, you can:
- Gain behavioral insight: Move beyond simple system metrics to understand if your agent is hallucinating, failing to follow conversation patterns, or losing grounding in its evidence.
- Automate alerts: Set up automated triggers that notify your team through tools like Slack when an agent's performance drops below a certain threshold.

> - Accelerate the research loop: Use the scores and failure analyses generated by signals to identify specific weaknesses, which can then be used to kick off the research loop for offline model improvement, data annotation, or reinforcement learning.

Reviewer: ```suggestion
- Accelerate the research loop: Use the scores and failure analyses generated by signals to identify specific weaknesses, which you can use to research model improvement, data annotation, or reinforcement learning.
```
"kick off" is kind of an idiom, so I try to avoid it.

## Available signals

> W&B Weave offers monitors with built-in signals: preset scorers that evaluate production traces for common quality issues and errors out of the box, with no custom setup. Each built-in signal uses a benchmarked LLM prompt to classify traces and saves the results as comma-delimited tags representing the detected issues.

Reviewer: I feel like this is more introductory stuff that should be at the top of the page.

> To start classifying traces immediately, enable signals from the Monitors page. Signals don't require prompt engineering or scorer configuration.

Reviewer: The workflows below guide people through this, so I don't think we need this line.

> Signals use a [W&B Inference](/inference/) model to score traces, so no external API keys are required.

Reviewer: Probably should either be in the intro area or the How signals work section. I'd opt for the introductory section and also add whether or not this costs customers additional money to use.

![Weave Traces view with a quality-classifier trace selected. The details panel shows Call details with classifiers metadata including a confidence score and reason.](/weave/guides/evaluation/img/weave_signals_trace_reasoning.png)

### See signals in the project dashboard
You can also review signals at a project level:
1. In the project sidebar, select **Project**.
1. At the top of the Project dashboard, select the **Weave** tab.
1. In the Weave dashboard panels, locate **Monitor Scores**.

In the Monitor Scores project panel, you can see time-based graphs of signals that occurred for the project. 
![Weave project dashboard Monitor Scores panel showing signal graphs from project activity.](/weave/guides/evaluation/img/weave_signals_project_dash.png)

### Alert on signals
To get notified when a signal is triggered, set up an [automation](/weave/guides/evaluation/automations).

## How signals work

Each signal uses an LLM-as-a-judge approach to classify traces:

> 1. **Trace selection**: Quality signals evaluate successful root-level traces. Error signals evaluate failed traces. Child spans and intermediate calls are not scored.

Reviewer: I feel like I'm missing context here. Are each of these a step that Weave goes through to assess and apply a signal to a trace?

Also, the "Trace selection" bullet introduces new information: quality signals only evaluate root-level traces. I feel like this should be mentioned earlier in "Available signals" so users don't wonder why child spans aren't scored.
